### ISSUE TITLE
Profiling slow page generation - wp-stateless accounts for over 37% of page generation time

### ISSUE THREAD

**Issue by ahmed-sigmalux** [udx/wp-stateless#265]
I'm profiling a website to figure out what's causing long (> 1s) page generation times. I've identified wp-stateless as one of the culprits, as it accounts for 16.76% of the cost. The problematic calls are:

```
wpCloud\StatelessMedia\Bootstrap->_convert_to_gs_link @ 654 - 10.74%

$meta[$key] = $this->_convert_to_gs_link($value, $image_host, $baseurl, $file_ext);
```

```
wpCloud\StatelessMedia\Bootstrap->convert_to_gs_link @ 636 - 6.02%

$updated = $this->_convert_to_gs_link($meta, $image_host, $baseurl, $file_ext);
```

Has anyone else profiled a website and identified the same issue? Any ideas why these 2 functions are slowing down page generation so much? And any ideas on how to speed them up, perhaps with object caching?

**Comment by ahmed-sigmalux** [comment]
And another 15.78% of the cost is:

```
wpCloud\StatelessMedia\Bootstrap->post_metadata_filter @ 609 - 13.66%

return $this->convert_to_gs_link(array_map('maybe_unserialize', $meta_cache[$meta_key]));
```

```
wpCloud\StatelessMedia\Bootstrap->post_metadata_filter @ 605 - 2.12%

return $this->convert_to_gs_link($meta_cache);
```

**Comment by ahmed-sigmalux** [comment]
And another 4.57% of the cost is:

```
wpCloud\StatelessMedia\Bootstrap->convert_to_gs_link @ 635 - 2.64%

$file_ext = $this->replaceable_file_types();
```

```
wpCloud\StatelessMedia\Bootstrap->convert_to_gs_link @ 634 - 1.93%

$types = $this->get('sm.body_rewrite_types');

```
So far wp-stateless accounts for over 37.11% of the page generation time!

**Comment by antonkorotkov** [comment]
@ahmed-sigmalux Try to disable option `File URL Replacement` and test again. In the meantime we will check why caching is not working in that place.

**Comment by alimuzzaman** [comment]
@antonkorotkov can you take a look at the `post_metadata_filter` and `convert_to_gs_link` function, if you can make anymore improvement.
I have added transient cache for one day in `post_metadata_filter`. But that could be bad idea, in  case of post updates.

**Comment by antonkorotkov** [comment]
@alimuzzaman I don't know how we can improve current implementation. I would think about re-implementing it somehow. For example save replaced URLs back to content so that it is processed only once. This is only an idea, not sure whether it will work or not.

**Comment by alimuzzaman** [comment]
@antonkorotkov This will be great for performance. We will only need to do it once.

- Convert local url to GCS url (and also sync the image).
- Save the changes to database.
- Create a meta to indicate that it's already processed.
- If the meta found we avoid the process.
- Hook into post save if any recent modification is made.
- We can add a sync option to make the changes at once instead of on the fly.

**Comment by ewsopp** [comment]
@antonkorotkov @alimuzzaman lets discuss this. We've discussed creating custom meta for attachments for several different needs. Before we start adding custom meta, I'd like to see a specification on our naming scheme and how it will be future proof.

**Comment by ewsopp** [comment]
@alimuzzaman is this related to the **File URL Replacement** setting?

**Comment by Mashenka** [comment]
@alimuzzaman can you answer Eric's question and provide steps how could I test that.

**Comment by alimuzzaman** [comment]
@ewsopp yes. Because we expanded to post meta it's time consuming.

We should exclude this from this milestone.
It is partially completed and there is more room for improvement.

**Comment by antonkorotkov** [comment]
@ewsopp 

I spent some time investigating the issue. I have some ideas.

So, yes, this issue is related to the option called **File URL Replacement**. 
The option description says `Scans post content and meta during presentation and replaces local media file urls with GCS urls.`

The issue is specifically related to the `meta` word in the description. There is no much trouble replacing post content (because it is single) but it is a problem to replace URLs in metadata since it is multiple and complex data that may be serialized or JSON-stringified etc. 

Metadata URLs replacement option was added as a `just-in-case` feature. We are not sure if it is useful for anyone at all. 

So, two ideas:
- we can separate this option out from **File URL Replacement** and make it disabled by default. It will still be available for those who really need it.
- we can simply remove `meta` replacing at all and it will fix the performance issue.

There are no much possibilities to improve the logic though because it is straightforward.

**Comment by ewsopp** [comment]
Agreed.

I created a separate issue for implementing these suggestions at https://github.com/wpCloud/wp-stateless/issues/336

**Comment by antonkorotkov** [comment]
Will close this once https://github.com/wpCloud/wp-stateless/issues/336 is closed.

**Comment by antonkorotkov** [comment]
Testing this: `ensure that "File URL Replacement" option is disabled by default for new installations.`

**Comment by antonkorotkov** [comment]
Ok, so to fix performance problem need to use `Disable` or `Enable Editor` options for `File URL Replacement` setting.

**Comment by tripledm** [comment]
hello, thanks for this thread super awesome! I have just one question if URL replacement set to Enable Editor, if it's run once on a post it's changed it and it won't need to change it again? Just to check sorry it might be an obvious question but it does state that it does not change the database and post content is stored there plus the permalink for a post so just checking. Thank you in advance.

**Comment by ewsopp** [comment]
No changes are made to the database.

If you want to make the change permanent, you can do a find/replace with other database related plugins or by working with the database directly (sequel pro, phpmyadmin, etc).

Hope that answers your question.
